### ISSUE TITLE
HTTP Server inference failed on Apple Silicon: "There is no Stream(gpu, 1) in current thread"

### ISSUE THREAD

**Issue by xjei75** [moona3k/mlx-qwen3-asr#16]
### Environment

- Device: MacBook Pro M4 (Apple Silicon, 128GB unified memory)
- macOS: 26.5.2
- Python: 3.12.13
- mlx-qwen3-asr: 0.3.5
- mlx: 0.32.0
- mlx-metal: 0.32.0

### Command

I started the OpenAI-compatible HTTP server:

```bash
mlx-qwen3-asr serve \
  --api-key <api-key> \
  --port 8765
```

The server starts successfully. However, when calling the transcription API:

```bash
curl http://localhost:8765/v1/audio/transcriptions \
  -H "Authorization: Bearer <api-key>" \
  -F "file=@Voice Memo.m4a" \
  -F "model=Qwen/Qwen3-ASR-0.6B"
```

the request reaches the server but fails with:

`{"error":{"message":"There is no Stream(gpu, 1) in current thread.","type":"server_error","param":null,"code":null}}`

I also verified that MLX works correctly outside the HTTP server:

```
python - <<'PY'
import mlx.core as mx
print(mx.default_device())
PY
```

Output:
`Device(gpu, 0)`

The issue only occurs when inference is triggered through the HTTP server. It looks like the inference execution path in the server may run in a different thread/context where the MLX GPU stream is not correctly initialized.
Could this be related to the async/thread execution model in the server implementation? Any suggestions for debugging or a workaround would be appreciated.

**Comment by Tadanobu0** [comment]
Reproduced on an M1 Max (macOS 14.7, Python 3.12, mlx 0.32.2 and 0.31.2, mlx-qwen3-asr 0.3.5). Your hunch about the thread model is right, and I opened #18 with a fix.

**Root cause.** MLX streams are thread-local. The server builds the `Session` in `lifespan` on the event-loop thread, but inference runs through `Session.transcribe_async`, which offloads to an arbitrary worker from `asyncio.to_thread`'s default pool. The failure lands at `mx.argmax(logits).item()` in `generate.py:502` — the first point where a value is pulled back into Python.

Both routes are affected (`/v1/audio/transcriptions` and the async `/transcribe`), which is why it looks like the whole server is broken rather than one endpoint.

**The non-obvious part.** Pinning inference to a dedicated thread is not enough — the model has to be *loaded* on that same thread. I measured all four combinations against a real model:

| model loaded on | inference runs on | result |
| --- | --- | --- |
| main thread | main thread (CLI) | OK |
| main thread | `asyncio.to_thread` | fails |
| main thread | dedicated single thread | fails |
| dedicated single thread | that same thread | OK |

That last row is the fix.

**Workaround until a fix lands.** The CLI path is unaffected, so shelling out to `mlx-qwen3-asr <file>` works today. If you need the server, the smallest local change is to make `lifespan` load the model inside a single-worker executor and run inference on that same executor:

```python
state.executor = ThreadPoolExecutor(max_workers=1, thread_name_prefix="mlx-asr")
loop = asyncio.get_running_loop()
state.session = await loop.run_in_executor(
    state.executor, lambda: Session(config.model, dtype=dtype)
)
```

then dispatch both call sites through `state.executor` instead of `transcribe_async`'s default pool. A single worker costs nothing here because the server already serializes inference behind `inference_lock`.

**Comment by moona3k** [comment]
Fixed in `main` (#19), thanks for the clear report and to @Tadanobu0 for #18.

Root cause: two arrays built during model construction (position-embedding table, MRoPE frequencies) were left as lazy graphs, and MLX 0.31+ binds those to the thread that built them. The server loaded the model on the event loop and ran inference on a pool thread, so the first forward pass from another thread failed. The library now materializes those buffers at construction, and the server additionally runs load and inference on one dedicated thread.

Verified on MLX 0.32.2 with a real model: both `/v1/audio/transcriptions` and the async `/transcribe` flow return transcripts. A PyPI release with this fix is not published yet; until then you can install from `main`:

```
pip install "git+https://github.com/moona3k/mlx-qwen3-asr@main#egg=mlx-qwen3-asr[serve]"
```

**Comment by moona3k** [comment]
Released in mlx-qwen3-asr 0.4.0 on PyPI: `pip install -U "mlx-qwen3-asr[serve]"`. The git install is no longer needed.
